frontend/components/SetupPage/FundYourAgent/hooks/useGetRefillRequirementsWithMonthlyGas.ts

```diff
@@ -0,0 +1,163 @@
+    } else {
+      const tokenDetails = getTokenDetails(
+        tokenAddress,
+        chainConfig,
+      ) as TokenConfig;
+      symbol = tokenDetails.symbol;
+      iconSrc = TokenSymbolConfigMap[tokenDetails.symbol as TokenSymbol]?.image;
+      decimals = tokenDetails.decimals;
+    }
+
+    if (symbol === TokenSymbolMap['XDAI']) {
+      iconSrc = '/tokens/wxdai-icon.png';
+    }
+
+    const parsedAmount = formatUnitsToNumber(amount, decimals);
+
+    if (parsedAmount > 0) {
+      tokenRequirements.push({
+        amount: parsedAmount,
+        symbol,
+        iconSrc,
+      });
+    }
+  });
+
+  return tokenRequirements.sort((a, b) => b.amount - a.amount);
+};
+
+export const useGetRefillRequimentsWithMonthlyGas = ({
+  selectedAgentConfig,
+  shouldCreateDummyService = false,
+}: {
+  selectedAgentConfig: AgentConfig;
+  shouldCreateDummyService?: boolean;
+}) => {
+  const updateBeforeBridgingFunds = useBeforeBridgeFunds();
+  const {
+    refillRequirements,
+    balances,
+    refetch,
+    isBalancesAndFundingRequirementsLoading,
+  } = useBalanceAndRefillRequirementsContext();
+  const { masterEoa } = useMasterWalletContext();
+
+  useEffect(() => {
+    const createDummyService = async () => {
+      await updateBeforeBridgingFunds();
+      await refetch?.();
+    };
+    if (shouldCreateDummyService) {
+      createDummyService();
+    }
+  }, [updateBeforeBridgingFunds, refetch, shouldCreateDummyService]);
+
+  const tokenRequirements = useMemo(() => {
+    if (!masterEoa || isBalancesAndFundingRequirementsLoading) return;
+
+    const { evmHomeChainId, middlewareHomeChainId } = selectedAgentConfig;
+    const chainConfig = TOKEN_CONFIG[evmHomeChainId];
+
+    // refill_requirements_masterEOA
+    const masterEoaRequirementAmount = (
+      refillRequirements as AddressBalanceRecord
+    )[masterEoa.address]?.[AddressZero];
+
+    // monthly_gas_estimate
+    const monthlyGasEstimate = BigInt(
+      SERVICE_TEMPLATES.find(
+        (template) => template.home_chain === middlewareHomeChainId,
+      )?.configurations[middlewareHomeChainId]?.monthly_gas_estimate ?? 0,
+    );
+
+    // native_token_amount in master_safe
+    const nativeTokenBalanceInMasterSafe = BigInt(
+      (balances as MasterSafeBalanceRecord)?.['master_safe']?.[AddressZero],
+    );
+
+    const requirementsPerToken = {} as { [tokenAddress: Address]: string };
+
+    Object.entries(
+      (refillRequirements as MasterSafeBalanceRecord)['master_safe'],
+    )?.forEach(([tokenAddress, amount]) => {
+      if (tokenAddress === AddressZero) {
+        /**
+         * If monthly gas estimate is greater than master_safe requirements, then consider that
+         * else, check the native tokens required to meet the gas requirements
+         */
+        const masterSafeRequirementAmount = BigInt(amount);
+        const gasDeficit = monthlyGasEstimate - nativeTokenBalanceInMasterSafe;
+        const amountNeededForGas = bigintMax(0n, gasDeficit);
```

Comment:
i thought this was needed given that monthly gas is handled client-side, but lemme know if we should ignore this. 
The `useGetBridgeRequirementsParamsWithMonthlyGasEstimate` hook currently doesn't take into account the balance on the master_safe

Reply:
Considering that if a user has some balance, it would be wrong if we don't take it into account, no? 🤔